backend/packager.go

```diff
@@ -0,0 +1,44 @@
+// Copyright Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/urlfetch"
+)
+
+func InitPackager() {
+	http.HandleFunc("/amppkg/", func(w http.ResponseWriter, r *http.Request) {
+		u, err := url.Parse(PACKAGER_PREFIX)
+		if err != nil {
+			log.Fatal(err)
+		}
+		p := httputil.NewSingleHostReverseProxy(u)
+		// Need to fiddle with Transport; on GAE this needs to change on every
+		// request. Might not be necessary under Go 1.11:
+		// https://cloud.google.com/blog/products/application-development/go-1-11-is-now-available-on-app-engine
+		p.Transport = &urlfetch.Transport{
+			Context: appengine.NewContext(r),
+		}
+		log.Printf("Proxying request for [%s] to [%s]", r.URL.String(), PACKAGER_PREFIX)
+		p.ServeHTTP(w, r)
+	})
+	http.HandleFunc("/priv/doc/", http.NotFound)
```

Comment:
why do we need this?

Reply:
Blocking access to `/priv/doc/` is one of the rules for [productionizing amppackager](https://github.com/ampproject/amppackager#productionizing)—`/priv/doc/` is supposed to be a "private" namespace used only for the front-end and `amppackager` to communicate.

Reply:
Isn't /priv/doc not handled by this server by default? That should be fine, I think. The important thing is that randos can't generate /priv/doc that are forwarded to the packager (without sanitization).

Reply:
There's a handler for `/` connected to the `/priv/doc/` proxying process so without blocking `/priv/doc/`, requests for `/priv/doc/` can end up hitting the packager (with appropriate request headers).

The URL gets a bit mangled (so a request for `https://ampbyexample.com/priv/doc/asdfasd` gets translated into `/priv/doc/?sign=https%3A%2F%2Fampbyexample.com%2Fpriv%2Fdoc%2Fasdfasd`) and so is difficult to exploit, but it seems worthwhile to just block here? I might be misunderstanding.

Reply:
Ah, I see! Well, I guess that prevents people from wasting ABE's time by constructing a really long recursive `/priv/doc` URL.